File: compilers/ruby/test/test_source_size_guard.rb
# frozen_string_literal: true

# BUG-008 follow-up: source-parser size-guard regression tests.

require_relative "test_helper"
require "runar_compiler/frontend/input_limits"
require "runar_compiler/compiler"

class TestSourceSizeGuard < Minitest::Test
  MAX = RunarCompiler::Frontend::InputLimits::MAX_SOURCE_BYTES
  ERR = RunarCompiler::Frontend::InputLimits::SourceSizeExceededError

  def test_assert_source_bytes_under_limit_rejects_oversized
    oversized = " " * (MAX + 1)
    err = assert_raises(ERR) do
      RunarCompiler::Frontend::InputLimits.assert_source_bytes_under_limit(oversized)
    end
    assert_equal MAX, err.limit
    assert_equal MAX + 1, err.actual
    assert_match(/MAX_SOURCE_BYTES/, err.message)
  end

  def test_assert_source_bytes_under_limit_accepts_at_limit
    sized = " " * MAX
    RunarCompiler::Frontend::InputLimits.assert_source_bytes_under_limit(sized)
  end

  def test_compiler_parse_source_rejects_oversized_across_all_extensions
    oversized = " " * (MAX + 1)
    %w[
      .runar.ts .runar.sol .runar.move .runar.go .runar.py
      .runar.rs .runar.rb .runar.zig .runar.java
    ].each do |ext|
      assert_raises(ERR) do
        RunarCompiler.send(:_parse_source, oversized, "Counter#{ext}")
      end
    end
  end

  def test_normal_source_does_not_trip_size_guard
    src = <<~RUBY
      require "runar"

      class Counter < RunarLang::SmartContract
        readonly :x
        def initialize(x); super; @x = x; end
        def unlock; end
      end
    RUBY
    # The Ruby parser may produce diagnostics; the only assertion is the
    # size guard did NOT raise.
    refute_raises(ERR) do
      begin
        RunarCompiler.send(:_parse_source, src, "Counter.runar.rb")
      rescue ERR
        raise
      rescue StandardError
        # other parse errors are fine
      end
    end
  end

  private

  def refute_raises(klass)
    yield
  rescue klass => e
    flunk "expected NOT to raise #{klass}, but did: #{e.message}"
  end
end

File: compilers/python/tests/test_source_size_guard.py
"""BUG-008 follow-up: source-parser size-guard regression tests."""

import pytest

from runar_compiler.frontend.input_limits import (
    MAX_SOURCE_BYTES,
    SourceSizeExceededError,
    assert_source_bytes_under_limit,
)
from runar_compiler.frontend.parser_dispatch import parse_source


def test_parse_source_rejects_oversized_input() -> None:
    oversized = " " * (MAX_SOURCE_BYTES + 1)
    with pytest.raises(SourceSizeExceededError) as exc:
        parse_source(oversized, "Counter.runar.ts")
    assert exc.value.limit == MAX_SOURCE_BYTES
    assert exc.value.actual == MAX_SOURCE_BYTES + 1


@pytest.mark.parametrize(
    "ext",
    [
        ".runar.ts",
        ".runar.sol",
        ".runar.move",
        ".runar.go",
        ".runar.py",
        ".runar.rs",
        ".runar.rb",
        ".runar.zig",
        ".runar.java",
    ],
)
def test_parse_source_rejects_oversized_input_regardless_of_extension(ext: str) -> None:
    oversized = " " * (MAX_SOURCE_BYTES + 1)
    with pytest.raises(SourceSizeExceededError):
        parse_source(oversized, "Counter" + ext)


def test_parse_source_accepts_normal_sized_input() -> None:
    # A minimal Python contract source. parse_source returns a ParseResult;
    # we only assert the size guard does not raise.
    src = (
        "from runar import SmartContract\n"
        "class Counter(SmartContract):\n"
        "    def __init__(self, x: int) -> None:\n"
        "        super().__init__()\n"
        "        self.x = x\n"
        "    @public\n"
        "    def unlock(self) -> None:\n"
        "        pass\n"
    )
    result = parse_source(src, "Counter.runar.py")
    # The Python parser may produce diagnostics; the only assertion is the
    # size guard did NOT raise.
    assert result is not None


def test_assert_source_bytes_under_limit_typed_error() -> None:
    with pytest.raises(SourceSizeExceededError) as exc:
        assert_source_bytes_under_limit(" " * (MAX_SOURCE_BYTES + 1))
    assert "MAX_SOURCE_BYTES" in str(exc.value)
    assert exc.value.limit == MAX_SOURCE_BYTES
